### ISSUE TITLE
transformer (forward pass)

### ISSUE THREAD

**Issue by slavavs82** [maxjcohen/transformer#10]
Parameters: | x (Tensor) – torch.Tensor of shape (batch_size, K, d_input).

What is K?

**Comment by maxjcohen** [comment]
Hi, `K` is the time dimension. In the case of the data challenge (see #2 ), we work on full months with hourly data, so `K = 24 * 28 = 672`. In the last examples on this repo, I've switch to weekly predictions, so `K = 24 * 7 = 168`.

**Comment by slavavs82** [comment]
Thank you so much for the answer!
I'll clarify, as my English prevents me from understanding correctly. K is the window size? d_input is the data size inside the window?

**Comment by maxjcohen** [comment]
No, `K` is the time dimension, for example the number of hours in `X`. `d_input` is the dimension of the input `X` for each time step. For instance, if you had 2 inputs (say outdoor temperature and ac schedule) for an entire week (168 hours), `X` shape would be `(1, 168, 2)`.

We refer to the window size as `"attention_size"`, see [here](https://timeseriestransformer.readthedocs.io/en/latest/transformer.html).

**Comment by slavavs82** [comment]
Fantastic, but I still don't get it :)
My data set is very simple (for the test). 
series = np.sin(np.arange(0, 1000)) 
I'm breaking this list down into 30 size windows.
I want to pass a window sized 30 to the network and predict the next 5 values. Let the Batch size be 1.
What size will x have? What would equal a K?

**Comment by maxjcohen** [comment]
In your case:
- `batch_size=1` : you only have one sample
- `K=1000` : 1000 time steps
- `d_input=1` : your sin is one dimensional

You don't have to break the list into size windows, the transformer will do it for you using the `score`, see [this line](https://github.com/maxjcohen/transformer/blob/master/tst/multiHeadAttention.py#L95) of the MHA block.

**Comment by slavavs82** [comment]
Dude, I really appreciate it.
Give me one last thing. How do I get a 30 size window online and predict the next 5?
I'm thinking logically. I have a data set. I want to take some data and teach the network to predict the next data that the network can't see.

**Comment by slavavs82** [comment]
I think I misled you myself.
series = np.sin(np.arange(0, 1000))
It's not a sample. It's a 1,000 size dataset.
I want to break this dataset down into samples. Size sample = 30. That's what I called a window. Then I can combine these samples into batch=4 (for example).
So, I have batch = 4, K = 30, d_input = 1. 
Is that so?

**Comment by maxjcohen** [comment]
Ok I think I see where you're going. In that case, your values for `batch_size`, `K` and `d_input` should be good. Unfortunately, the Transformer isn't well fit to predict future states, as it predicts one single output for each input. We discuss this in #5 .

**Comment by slavavs82** [comment]
In this case, there is a question.
I submit sample=30(values) to the encoder input. If I want to predict the next 10 values, I submit these 10 values to the decoder input. I have seen this scheme in this document https://arxiv.org/pdf/2001.08317.pdf. Look at figure 1. 
Encoder input = T1,T2,T3,T4; decoder input T4,T5. In my case I want to feed Encoder input = T1,T2...T30; decoder input T30,T31...T40.
Can I do that?

**Comment by pqy000** [comment]
Of course, you can try it ?

**Comment by Lisa-FFY** [comment]
Excuse me,could you please tell me is there any relevant paper about this code?I want to study it in depth.

**Comment by maxjcohen** [comment]
Hi, this seems to be an issue with Pandas, as you can see in the last stack of the Traceback. Did you try to feed a pandas Dataframe directly to the trainer ? You most likely need to modify the dataloader class to match your dataset.

In the future, please open a new issue when discussing new/different problems. Thanks !
